### PR TITLE
Fix race condition in ProfilesCache refresh

### LINKED ISSUE #2831: ProfilesCache does not handle race conditions
**Describe the bug**

Although JavaScript and Node.js are single-threaded, multiple VS Code extensions can activate in parallel. Extensions that grab the static `ProfilesCache` instance from Zowe Explorer might compete with one another when performing operations on the cache (such as `refresh`), causing undefined behavior in the cache (such as missing or incorrect profile values, etc.)

**To Reproduce**

Have two or more extensions that access the `ProfilesCache` instance through Zowe Explorer's extension exports (`vscode.extensions.getExtension("zowe.vscode-extension-for-zowe").exports`). Try to call `refresh` on this cache instance for each extension, as soon as the extensions are activated.

**Expected behavior**

The ProfilesCache instance should fill the cache successfully during refresh, and return an instance with valid profiles.

**Desktop (please complete the following information):**

- OS: Windows
- Zowe Explorer Version: 2.15.2

**Additional context**

A few internal developers have reported this issue. Due to the nature of VS Code extensions and asynchronous code in Node.js, we likely need to add multi-threading support to prevent any problems that may surface from ProfilesCache race conditions.

### PR DESCRIPTION
**Pull request by t1m0thyj**
## Proposed changes

Fixes #2831

## Release Notes

Milestone: 2.15.3

Changelog: Fixed an issue where `ProfilesCache` may return missing or incorrect profile values when multiple extensions call it during activation

## Types of changes

What types of changes does your code introduce to Zowe Explorer?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Updates to Documentation or Tests (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This checklist will be used as reference for both the contributor and the reviewer_

- [ ] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/zowe-explorer-vscode/wiki/Best-Practices:-Contributor-Guidance) wiki
- [ ] PR title follows [Conventional Commits Guidelines](https://www.conventionalcommits.org/en/v1.0.0-beta.2/)
- [ ] PR Description is included
- [ ] gif or screenshot is included if visual changes are made
- [ ] `yarn workspace vscode-extension-for-zowe vscode:prepublish` has been executed
- [ ] All checks have passed (DCO, Jenkins and Code Coverage)
- [ ] I have added unit test and it is passing
- [ ] I have added integration test and it is passing
- [ ] There is coverage for the code that I have added
- [ ] I have tested it manually and there are no regressions found
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any PR dependencies have been merged and published (if appropriate)

### CODE REVIEW

File: packages/zowe-explorer/package.json [hunk 1]
```diff
@@ -823,7 +823,7 @@
       },
       {
         "command": "zowe.placeholderCommand",
-        "title": "%zowe.placeholderCommand%",
+        "title": "",
         "enablement": "false"
       }
```
Comment (zFernand0) [inline packages/zowe-explorer/package.json:828]: Dumb questions: 😋 
- What's this?
- Can we remove it?

File: packages/zowe-explorer-api/src/profiles/ProfilesCache.ts [hunk 2]
```diff
@@ -219,10 +216,15 @@ export class ProfilesCache {
...
+                    allProfiles.push(...tmpAllProfiles);
                     this.profilesByType.set(type, tmpAllProfiles);
                 }
-                this.allTypes.push(type);
+            }
+            this.allProfiles = allProfiles;
+            this.allTypes = allTypes;
+            for (const oldType of [...this.profilesByType.keys()].filter((type) => !allTypes.includes(type))) {
+                this.profilesByType.delete(oldType);
+                this.defaultProfileByType.delete(oldType);
```
Comment (zFernand0) [inline packages/zowe-explorer-api/src/profiles/ProfilesCache.ts:227]: I know this for-loop is based on a filtered version of `this.profileByType`, so doing `this.profileByType.delete(oldType)` should be fine.
I'm curious if we should check if `this.defaultProfileByType.has(oldType)` before trying to delete it. 😋

**Review by zFernand0 — APPROVED** [review]
Thanks for attempting to fix the theia tests. 😋 
Also, I can see how this is attempting to resolve the race condition and avoiding missing profile values (by not clearing `this.allTypes` on every refresh), but if you could provide some quick step to test how to get incorrect profile values, or maybe show this on a future standup?

Changes LGTM though! 😋

File: packages/zowe-explorer-api/src/profiles/ProfilesCache.ts (hunk 2 above)
Comment (traeok) [reply packages/zowe-explorer-api/src/profiles/ProfilesCache.ts:227]: The current logic should be okay, as `delete` will [return false](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/delete#return_value) if the element doesn't exist

File: packages/zowe-explorer/package.json (hunk 1 above)
Comment (t1m0thyj) [reply packages/zowe-explorer/package.json:828]: We use this to show tree nodes as disabled. VS Code will "gray out" a tree node that has a command assigned with "enablement = false". Thanks @traeok for suggesting this trick 😋

File: packages/zowe-explorer/package.json [hunk 3]
```diff
@@ -823,7 +823,7 @@
       },
       {
         "command": "zowe.placeholderCommand",
-        "title": "%zowe.placeholderCommand%",
+        "title": "",
```
Comment (traeok) [inline packages/zowe-explorer/package.json:826]: We'll probably want to switch back to a localized placeholder, as a side effect of the empty `title` was found while addressing feedback in #2777:

image

_(only affects extension host)_

**Review by traeok — COMMENTED** [review]
LGTM, thanks @t1m0thyj for resolving this issue! I left a comment regarding the placeholder command based on our previous offline discussions

**Review by traeok — APPROVED** [review]

**Review by JillieBeanSim — APPROVED** [review]
LGTM! thanks @t1m0thyj

**Comment by traeok** [comment]
Hi @roman-kupriyanov,

The second block of code you posted would be the right approach with the current state of the ProfilesCache. In `initForZowe` we do a `refresh` of the profiles cache, and during that refresh, we get all the types using `ProfilesCache.getAllProfileTypes`. If you do not call `registerCustomProfilesType` beforehand, it will not consider that type unless it has already been registered under one of the APIs (MVS, USS, or JES).

Ideally, I'd like to register those custom types as a part of the logic in `initForZowe` so that extenders don't have to call `registerCustomProfilesType` - rather, we can do it ourselves during initialization. I don't think this would be breaking since `allExternalTypes` is already a set, so continuing to call the `registerCustomProfilesType` function shouldn't affect anything.

I don't want to speak for @t1m0thyj, but those are my thoughts regarding extender initialization/preparing custom profile types 😋

**Comment by roman-kupriyanov** [comment]
Hi @traeok,

Thanks for the quick response!

> I believe the second block of code you posted would be the right approach with the current state of the ProfilesCache. In `initForZowe` we do a `refresh` of the profiles cache, and during that refresh, we get all the types using `ProfilesCache.getAllProfileTypes`. If you do not call `registerCustomProfilesType` beforehand, it will not consider that type unless it has already been registered under one of the APIs (MVS, USS, or JES).

That was my assumption too since it makes total sense to have an external profile type registered before the attempt of refreshing the cache. I think, before that extra `reloadProfiles()` call was actually protecting us from the problems by refreshing one more time, when all the parameters are actually set.

> Ideally, I'd like to register those custom types as a part of the logic in `initForZowe` so that extenders don't have to call `registerCustomProfilesType` - rather, we can do it ourselves during initialization if the type doesn't already exist. I don't think this would be breaking since `allExternalTypes` is already a set, so continuing to call the `registerCustomProfilesType` function from an extender's perspective shouldn't affect anything.

Those were my thoughts exactly! Moreover, if the order of those calls is crucial to make it work right, I would expect it to be called internally by `initForZowe()` and at the right time to avoid making a mistake as the extender.

But I am not going to force any decision here as well and leave it up to @t1m0thyj  and the team. 

Thanks for the confirmation and investigation, cheers!

**Comment by JillieBeanSim** [comment]
Thanks for testing the different scenarios of registering with ZE @roman-kupriyanov.  We will want to make sure our documentation matches the best solution for registering with ZE in zowe.org and our wiki if it's different than what we already have so other extenders can benefit from it. ❤️

**Review by awharn — APPROVED** [review]

**Pull request merged**